packages/apps/plugins/plugin-debug/src/scaffolding/generator.ts

```diff
@@ -0,0 +1,131 @@
+  randomText,
+  SpaceObjectGenerator,
+  type TestGeneratorMap,
+  type TestMutationsMap,
+  type TestSchemaMap,
+} from '@dxos/echo-generator';
+import { create } from '@dxos/echo-schema';
+import { faker } from '@dxos/random';
+
+export enum SchemasNames {
+  document = 'dxos.org/type/Document',
+  diagram = 'dxos.org/type/Diagram',
+}
+
+export const SchemasMap: TestSchemaMap<SchemasNames> = {
+  [SchemasNames.document]: DocumentType,
+  [SchemasNames.diagram]: DiagramType,
+};
+
+export const ObjectGenerators: TestGeneratorMap<SchemasNames> = {
+  [SchemasNames.document]: () => {
+    const name = faker.lorem.sentence();
+
+    return { name, content: create(TextType, { content: '' }), threads: [] };
+  },
+
+  [SchemasNames.diagram]: () => {
+    const name = faker.lorem.sentence();
+    return {
+      name,
+      canvas: create(CanvasType, { schema: TLDRAW_SCHEMA, content: {} }),
+    };
+  },
+};
+
+export const MutationsGenerators: TestMutationsMap<SchemasNames> = {
+  [SchemasNames.document]: (object, params) => {
+    const accessor = createDocAccessor(object.content, ['content']);
+
+    for (let i = 0; i < params.count; i++) {
+      const length = object.content?.content?.length ?? 0;
+      accessor.handle.change((doc) => {
+        A.splice(
+          doc,
+          accessor.path.slice(),
+          0,
+          params.maxContentLength >= length ? 0 : params.mutationSize,
+          randomText(params.mutationSize) + ' ',
+        );
+      });
+    }
+  },
+
+  [SchemasNames.diagram]: (object, params) => {
+    const store = new AutomergeStoreAdapter({ timeout: 250 });
+    store.open(createDocAccessor(object.canvas, ['content']));
+    const app = new Editor({
+      store: store.store!,
+      shapeUtils: defaultShapeUtils,
+      bindingUtils: defaultBindingUtils,
+      tools: defaultTools,
+      getContainer: () => document.getElementsByTagName('body')[0],
+    });
+
+    //
+    // Draw spiral.
+    //
+    const r = 100;
+    const a = 0.05;
+    const cx = 200;
+    const cy = 200;
+
+    for (let i = 0; i < params.count; i++) {
+      const t = i;
+      const t1 = i + 1;
+      const x = cx + a * t * r * Math.cos(t);
+      const y = cy + a * t * r * Math.sin(t);
+      const x1 = cx + a * t1 * r * Math.cos(t1);
+      const y1 = cy + a * t1 * r * Math.sin(t1);
+
+      app.createShape({
+        id: createShapeId(),
+        isLocked: false,
+        meta: {},
+        opacity: 1,
+        parentId: 'page:page' as TLParentId,
+        props: {
+          arrowheadEnd: 'none',
+          arrowheadStart: 'none',
+          bend: 0,
+          color: 'black',
+          dash: 'draw',
+          start: { x, y },
+          end: { x: x1, y: y1 },
+          fill: 'none',
+          font: 'draw',
+          labelColor: 'black',
+          labelPosition: 0.5,
+          scale: 1,
+          size: 'm',
+        },
+        rotation: 0,
+        type: 'arrow',
+        typeName: 'shape',
+        x: 0,
+        y: 0,
```

Comment:
does this always draw it in the same place? maybe helpful to place them at random locations?

Reply:
See video

Reply:
Right but that looks like it's only drawing one spiral per diagram? what happens if more generated, are they just drawn on top or at a new start point?

Reply:
I do not get the question, it draws one spiral per diagram, the spiral size depends on the number of mutations, each mutation is a separate segment of the spiral, and currently, there is no way to mutate the diagram a second time. So it doesn't matter what starting coordinates we choose

Reply:
Ah I didn't realize that you couldn't mutate the diagram a second time. I though that the mutator chose a random object to mutate each time. Makes sense 👍